src/Std/Time/Format.lean

```diff
@@ -0,0 +1,437 @@
+-/
+def toDateTimeWithZoneString (pdt : ZonedDateTime) : String :=
+  Formats.dateTimeWithZone.format pdt.snd
+
+/--
+Parses a `String` in the `ISO8601`, `RFC822` or `RFC850` format and returns a `ZonedDateTime`.
+-/
+def parse (input : String) : Except String ZonedDateTime :=
+  fromISO8601String input
+  <|> fromRFC822String input
+  <|> fromRFC850String input
+
+instance : ToString ZonedDateTime where
+  toString := toDateTimeWithZoneString
+
+instance : Repr ZonedDateTime where
+  reprPrec data := Repr.addAppParen (toString data)
+
+end ZonedDateTime
+
+namespace PlainDateTime
+
+/--
+Formats a `PlainDateTime` using a specific format.
+-/
+def format (date : PlainDateTime) (format : String) : String :=
+  let format : Except String (Format .any) := Format.spec format
+  match format with
+  | .error err => s!"error: {err}"
+  | .ok res =>
+    let res := res.formatGeneric fun
+      | .YYYY | .YY => some date.year
+      | .MMMM | .MMM | .MM | .M => some date.month
+      | .DD | .D | .d => some date.day
+      | .EEEE | .EEE => some date.date.weekday
+      | .HH | .H => some date.time.hour
+      | .mm | .m => some date.time.minute
+      | .sss => some (Internal.Bounded.LE.ofNat 0 (by decide))
+      | .ss | .s => some date.time.second
+      | _ => none
+    match res with
+    | some res => res
+    | none => "invalid time"
+
+/--
+Parses a `String` in the `AscTime` format and returns a `PlainDateTime` object in the GMT time zone.
+-/
+def fromAscTimeString (input : String) : Except String PlainDateTime :=
+  Formats.ascTime.parse input
+  |>.map DateTime.toPlainDateTime
+
+/--
+Formats a `PlainDateTime` value into an AscTime format string.
+-/
+def toAscTimeString (pdt : PlainDateTime) : String :=
+  Formats.ascTime.format (DateTime.ofPlainDateTime pdt .UTC)
+
+/--
+Parses a `String` in the `LongDateFormat` and returns a `PlainDateTime` object in the GMT time zone.
+-/
+def fromLongDateFormatString (input : String) : Except String PlainDateTime :=
+  Formats.longDateFormat.parse input
+  |>.map DateTime.toPlainDateTime
+
+/--
+Formats a `PlainDateTime` value into a LongDateFormat string.
+-/
+def toLongDateFormatString (pdt : PlainDateTime) : String :=
+  Formats.longDateFormat.format (DateTime.ofPlainDateTime pdt .UTC)
+
+/--
+Parses a `String` in the `DateTime` format and returns a `PlainDateTime`.
+-/
+def fromDateTimeString (input : String) : Except String PlainDateTime :=
+  Formats.dateTime24Hour.parse input
+  |>.map DateTime.toPlainDateTime
+
+/--
+Formats a `PlainDateTime` value into a `DateTime` format string.
+-/
+def toDateTimeString (pdt : PlainDateTime) : String :=
+  Formats.dateTime24Hour.format (DateTime.ofPlainDateTime pdt .UTC)
+
+/--
+Parses a `String` in the `AscTime` or `LongDate` format and returns a `PlainDateTime`.
+-/
+def parse (date : String) : Except String PlainDateTime :=
+  fromAscTimeString date
+  <|> fromLongDateFormatString date
+
+instance : ToString PlainDateTime where
```

Comment:
In practical usage I found it quite frustrating that this doesn't give the full precision that the type offers (i.e., nanoseconds). This is certainly not acceptable for the `Repr` instance, because that needs to round-trip, so it needs to return something (like a `date%` literal) that Lean can parse into an equal object.

Reply:
I added the full precision to each type and a form that can be parsed by the `date%`.